crates/services/gas_price_service/src/v1/service.rs

```diff
@@ -0,0 +1,473 @@
+
+    impl FakeMetadata {
+        fn empty() -> Self {
+            Self {
+                inner: Arc::new(std::sync::Mutex::new(None)),
+            }
+        }
+    }
+
+    impl MetadataStorage for FakeMetadata {
+        fn get_metadata(
+            &self,
+            _: &BlockHeight,
+        ) -> GasPriceResult<Option<UpdaterMetadata>> {
+            let metadata = self.inner.lock().unwrap().clone();
+            Ok(metadata)
+        }
+
+        fn set_metadata(&mut self, metadata: &UpdaterMetadata) -> GasPriceResult<()> {
+            *self.inner.lock().unwrap() = Some(metadata.clone());
+            Ok(())
+        }
+    }
+
+    #[tokio::test]
+    async fn run__updates_gas_price_with_l2_block_source() {
+        // given
+        let block_height = 1;
+        let l2_block = BlockInfo::Block {
+            height: block_height,
+            gas_used: 60,
+            block_gas_capacity: 100,
+        };
+
+        let (l2_block_sender, l2_block_receiver) = mpsc::channel(1);
+        let l2_block_source = FakeL2BlockSource {
+            l2_block: l2_block_receiver,
+        };
+
+        let metadata_storage = FakeMetadata::empty();
+        let l2_block_height = 0;
+        let config = V1AlgorithmConfig {
+            new_exec_gas_price: 100,
+            min_exec_gas_price: 50,
+            exec_gas_price_change_percent: 20,
+            l2_block_fullness_threshold_percent: 20,
+            gas_price_factor: NonZeroU64::new(10).unwrap(),
+            min_da_gas_price: 10,
+            max_da_gas_price_change_percent: 20,
+            da_p_component: 4,
+            da_d_component: 2,
+            normal_range_size: 10,
+            capped_range_size: 100,
+            decrease_range_size: 4,
+            block_activity_threshold: 20,
+        };
+        let (algo_updater, shared_algo) =
+            initialize_algorithm(&config, l2_block_height, &metadata_storage).unwrap();
+
+        let notifier = Arc::new(tokio::sync::Notify::new());
+        let dummy_da_source = DaSourceService::new(
+            DummyDaBlockCosts::new(
+                Err(anyhow::anyhow!("unused at the moment")),
+                notifier.clone(),
+            ),
+            None,
+        );
+
+        let mut service = GasPriceServiceV1::new(
+            l2_block_source,
+            metadata_storage,
+            shared_algo,
+            algo_updater,
+            dummy_da_source,
+        );
+        let read_algo = service.next_block_algorithm();
+        let mut watcher = StateWatcher::default();
+        let initial_price = read_algo.next_gas_price();
+
+        // when
+        service.run(&mut watcher).await.unwrap();
+        l2_block_sender.send(l2_block).await.unwrap();
+        service.shutdown().await.unwrap();
+
+        // then
+        let actual_price = read_algo.next_gas_price();
+        assert_ne!(initial_price, actual_price);
+    }
+
+    #[tokio::test]
+    async fn run__updates_gas_price_with_da_block_cost_source() {
```

Comment:
cc: @MitchTurner this test seems to be failing with the following error - 
```
called `Result::unwrap()` on an `Err` value: L2 block expected but not found in unrecorded blocks: 1
```

which i narrowed down to https://github.com/FuelLabs/fuel-core/blob/a14922eb17a9ce844132edc240b5157eb860199f/crates/fuel-gas-price-algorithm/src/v1.rs#L546-L562.

This is what happens when the da block cost that is received refers to an L2 block that hasn't been processed by the algorithm yet. that is fine, seems like defined behaviour.


Below, on line 432-434, I define the cost and bytes. Something seems to be wrong in the `drain_l2_block_bytes_for_range` function https://github.com/FuelLabs/fuel-core/blob/a14922eb17a9ce844132edc240b5157eb860199f/crates/fuel-gas-price-algorithm/src/v1.rs#L546 when I call it with just one item in the range. because the `range_bytes` it yields after being called in https://github.com/FuelLabs/fuel-core/blob/a14922eb17a9ce844132edc240b5157eb860199f/crates/fuel-gas-price-algorithm/src/v1.rs#L529 is `0`, which doesn't seem too correct.

Reply:
wondering if you have any thoughts about the params I'm using.

Reply:
> wondering if you have any thoughts about the params I'm using.

With these tests, the values don't exactly matter. We use an `assert_ne!` and as long as that's true we're not concerned with the details. The problem with this approach is the test writer/reader still needs to know what values would result in a different gas price. This is part of the reason I originally had an abstraction of gas price updater (which we've (you've) removed)